### PR TITLE
Changed book repository, added new methods and endpoints in bookServi…

### PR DESCRIPTION
**Pull request by f3ops3**
…ce and BookController, implemented soft delete concept, added liquibase support.

### CODE REVIEW

File: src/main/java/springweb/courseproject/repository/BookRepository.java [hunk 1]
```diff
@@ -1,13 +1,10 @@
...
 
-public interface BookRepository {
-    Book save(Book book);
-
-    List<Book> findAll();
-
-    Optional<Book> findById(long id);
+@Repository
+public interface BookRepository extends JpaRepository<Book, Long> {
+    //Book updateBookById(Long id, Book book);
```
Comment (Elena-Bruyako) [inline src/main/java/springweb/courseproject/repository/BookRepository.java:9]: remove commented code

File: src/main/java/springweb/courseproject/service/BookService.java [hunk 2]
```diff
@@ -10,4 +10,8 @@ public interface BookService {
     List<BookDto> findAll();
 
     BookDto findById(long id);
+
+    BookDto updateBookById(long id, CreateBookRequestDto book);
```
Comment (Elena-Bruyako) [inline src/main/java/springweb/courseproject/service/BookService.java:14]:
```suggestion
    BookDto updateBookById(Long id, CreateBookRequestDto book);

```

File: src/main/java/springweb/courseproject/service/BookService.java [hunk 3]
```diff
@@ -10,4 +10,8 @@ public interface BookService {
     List<BookDto> findAll();
 
     BookDto findById(long id);
+
+    BookDto updateBookById(long id, CreateBookRequestDto book);
+
+    void deleteBookById(long id);
```
Comment (Elena-Bruyako) [inline src/main/java/springweb/courseproject/service/BookService.java:16]:
```suggestion
    void deleteBookById(Long id);
```

File: src/main/java/springweb/courseproject/repository/BookRepository.java [hunk 4]
```diff
@@ -1,13 +1,10 @@
...
+import org.springframework.stereotype.Repository;
 import springweb.courseproject.model.Book;
 
-public interface BookRepository {
-    Book save(Book book);
-
-    List<Book> findAll();
-
-    Optional<Book> findById(long id);
+@Repository
```
Comment (Elena-Bruyako) [inline src/main/java/springweb/courseproject/repository/BookRepository.java:7]:
```suggestion
```

File: src/main/java/springweb/courseproject/service/BookServiceImpl.java [hunk 5]
```diff
@@ -35,4 +35,19 @@ public BookDto findById(long id) {
                 () -> new EntityNotFoundException("Couldn't find book with id = " + id))
         );
     }
+
+    @Override
+    public BookDto updateBookById(long id, CreateBookRequestDto book) {
+        Book bookToUpdate = bookRepository.findById(id).orElseThrow(
+                () -> new EntityNotFoundException("Couldn't find book with id = " + id));
+        bookToUpdate = bookMapper.updateBook(book, bookToUpdate);
```
Comment (Elena-Bruyako) [inline src/main/java/springweb/courseproject/service/BookServiceImpl.java:43]:
```suggestion
        Book book = bookRepository.findById(id).orElseThrow(
                () -> new EntityNotFoundException("Couldn't find book with id = " + id));
        bookMapper.updateBook(bookDto, book);
```

File: src/main/java/springweb/courseproject/service/BookServiceImpl.java [hunk 6]
```diff
@@ -35,4 +35,19 @@ public BookDto findById(long id) {
                 () -> new EntityNotFoundException("Couldn't find book with id = " + id))
         );
     }
+
+    @Override
+    public BookDto updateBookById(long id, CreateBookRequestDto book) {
```
Comment (Elena-Bruyako) [inline src/main/java/springweb/courseproject/service/BookServiceImpl.java:40]:
```suggestion
    public BookDto updateBookById(long id, CreateBookRequestDto bookDto) {
```

File: src/main/java/springweb/courseproject/service/BookServiceImpl.java [hunk 7]
```diff
@@ -35,4 +35,19 @@ public BookDto findById(long id) {
                 () -> new EntityNotFoundException("Couldn't find book with id = " + id))
         );
     }
+
+    @Override
+    public BookDto updateBookById(long id, CreateBookRequestDto book) {
+        Book bookToUpdate = bookRepository.findById(id).orElseThrow(
+                () -> new EntityNotFoundException("Couldn't find book with id = " + id));
+        bookToUpdate = bookMapper.updateBook(book, bookToUpdate);
+        return bookMapper.toDto(bookRepository.save(bookToUpdate));
```
Comment (Elena-Bruyako) [inline src/main/java/springweb/courseproject/service/BookServiceImpl.java:44]:
```suggestion
        return bookMapper.toDto(bookRepository.save(book));

```

File: src/main/java/springweb/courseproject/mapper/BookMapper.java [hunk 8]
```diff
@@ -11,4 +12,6 @@ public interface BookMapper {
     BookDto toDto(Book book);
 
     Book toBook(CreateBookRequestDto dto);
+
+    Book updateBook(CreateBookRequestDto dto, @MappingTarget Book book);
```
Comment (Elena-Bruyako) [inline src/main/java/springweb/courseproject/mapper/BookMapper.java:16]:
```suggestion
    void updateBook(CreateBookRequestDto bookDto, @MappingTarget Book book);
```

File: src/main/java/springweb/courseproject/service/BookServiceImpl.java [hunk 9]
```diff
@@ -35,4 +35,19 @@ public BookDto findById(long id) {
...
+    public BookDto updateBookById(long id, CreateBookRequestDto book) {
+        Book bookToUpdate = bookRepository.findById(id).orElseThrow(
+                () -> new EntityNotFoundException("Couldn't find book with id = " + id));
+        bookToUpdate = bookMapper.updateBook(book, bookToUpdate);
+        return bookMapper.toDto(bookRepository.save(bookToUpdate));
+    }
+
+    @Override
+    public void deleteBookById(long id) {
+        Book bookToDelete = bookRepository.findById(id).orElseThrow(
```
Comment (Elena-Bruyako) [inline src/main/java/springweb/courseproject/service/BookServiceImpl.java:49]: if there are no requirements to throw an exception during delete, this request to DB could be redundant.

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve it

File: src/main/java/springweb/courseproject/service/BookServiceImpl.java [hunk 10]
```diff
@@ -30,9 +30,22 @@ public List<BookDto> findAll() {
...
+    public BookDto updateBookById(Long id, CreateBookRequestDto bookDto) {
+        Book book = bookRepository.findById(id).orElseThrow(
+                () -> new EntityNotFoundException("Couldn't find book with id = " + id));
+        bookMapper.updateBook(bookDto, book);
+        return bookMapper.toDto(bookRepository.save(book));
+    }
+
+    @Override
+    public void deleteBookById(Long id) {
+        bookRepository.findById(id).ifPresent(bookRepository::delete);
```
Comment (Elena-Bruyako) [inline src/main/java/springweb/courseproject/service/BookServiceImpl.java:49]:
```suggestion
        bookRepository.findById(id);

```

File: src/main/resources/db/changelog/changes/01-create-books-table.yaml [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+                  name: description
+                  type: text
+              - column:
+                  name: cover_image
+                  type: varchar(255)
+              - column:
+                  name: is_deleted
+                  type: tinyint(1)
+                  constraints:
+                    nullable: false
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/01-create-books-table.yaml:47]: add an empty line at the end of file

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by MentorMateAcademy — APPROVED** [review]

**Pull request merged**